reactor-core/src/main/java/reactor/util/retry/RetryBuilder.java

```diff
@@ -0,0 +1,375 @@
+			BiFunction<RetrySignal, Mono<Void>, Mono<Void>> asyncPostRetry,
+			BiFunction<RetryBuilder, RetrySignal, Throwable> retryExhaustedGenerator) {
+		this.maxAttempts = max;
+		this.throwablePredicate = aThrowablePredicate::test; //massaging type
+		this.isTransientErrors = isTransientErrors;
+		this.doPreRetry = doPreRetry;
+		this.doPostRetry = doPostRetry;
+		this.asyncPreRetry = asyncPreRetry;
+		this.asyncPostRetry = asyncPostRetry;
+		this.retryExhaustedGenerator = retryExhaustedGenerator;
+	}
+
+	/**
+	 * Set the maximum number of retry attempts allowed. 1 meaning "1 retry attempt":
+	 * the original subscription plus an extra re-subscription in case of an error, but
+	 * no more.
+	 *
+	 * @param maxAttempts the new retry attempt limit
+	 * @return the builder for further configuration
+	 */
+	public RetryBuilder maxAttempts(long maxAttempts) {
+		return new RetryBuilder(
+				maxAttempts,
+				this.throwablePredicate,
+				this.isTransientErrors,
+				this.doPreRetry,
+				this.doPostRetry,
+				this.asyncPreRetry,
+				this.asyncPostRetry,
+				this.retryExhaustedGenerator);
+	}
+
+	/**
+	 * Set the {@link Predicate} that will filter which errors can be retried. Exceptions
+	 * that don't pass the predicate will be propagated downstream and terminate the retry
+	 * sequence. Defaults to allowing retries for all exceptions.
+	 *
+	 * @param predicate the predicate to filter which exceptions can be retried
+	 * @return the builder for further configuration
+	 */
+	public RetryBuilder throwablePredicate(Predicate<? super Throwable> predicate) {
+		return new RetryBuilder(
+				this.maxAttempts,
+				Objects.requireNonNull(predicate, "predicate"),
+				this.isTransientErrors,
+				this.doPreRetry,
+				this.doPostRetry,
+				this.asyncPreRetry,
+				this.asyncPostRetry,
+				this.retryExhaustedGenerator);
+	}
+
+	/**
+	 * Allows to augment a previously {@link #throwablePredicate(Predicate) set} {@link Predicate} with
+	 * a new condition to allow retries of some exception or not. This can typically be used with
+	 * {@link Predicate#and(Predicate)} to combine existing predicate(s) with a new one.
+	 * <p>
+	 * For example:
+	 * <pre><code>
+	 * //given
+	 * RetryBuilder retryTwiceIllegalArgument = Retry.max(2)
+	 *     .throwablePredicate(e -> e instanceof IllegalArgumentException);
+	 *
+	 * RetryBuilder retryTwiceIllegalArgWithCause = retryTwiceIllegalArgument.throwablePredicate(old ->
+	 *     old.and(e -> e.getCause() != null));
+	 * </code></pre>
+	 *
+	 * @param predicateAdjuster a {@link Function} that returns a new {@link Predicate} given the
+	 * currently in place {@link Predicate} (usually deriving from the old predicate).
+	 * @return a new copy of the builder which can either be further configured or used as {@link Retry}
+	 */
+	public RetryBuilder throwablePredicateModifiedWith(
+			Function<Predicate<Throwable>, Predicate<? super Throwable>> predicateAdjuster) {
```

Comment:
yes, probably one can only use `and`. But the `filter` method that replaces the `Predicate` is useful when one wants to change the predicate from a template kind of spec that was shared. Maybe reverse the logic and have the second method be the one that forces a replace (eg. `filter` would use `and` and `replaceFilter` would just replace)? this could add some consistency with the fact that the `doBefore`/`doAfter` are all additive

Reply:
That sounds good to me.